### PR TITLE
Solr plugin impl

### PR DESCRIPTION
**Pull request by aperullo**
### Description of the Change

This PR represents a new maven plugin to replace the current way the Solr database is acquired and started for the integration tests. The current implementation uses the exec maven plugin to directly execute the Solr bash/cmd script, this new plugin wraps that functionality and checks the success of the commands.

There are three goals within the plugin.
1. **Prepare**: Downloads the Solr-distro dependency if it is not available locally, and unpacks it into the target directory.
2. **Start**: After figuring out which OS the plugin is running on, this goal will invoke the start command with the specified port (8983 by default). If Solr fails to start, the build will be halted with an error and the cause will be inserted into the debug level for investigation. 
3. **Stop**: After the integration tests have finished running, and after figuring out which OS the plugin is running on, this goal will invoke the stop command with the specified port. If Solr fails to stop, the user will be warned and it will try to stop Solr again. If this fails then an error will be printed to the log; however, the build will not be halted as the integration tests have already passed.
(The built-in test for this functionality will only work following [the change for line endings on windows](https://codice.atlassian.net/browse/DDF-4432)) This goal also should be set up to run both before and after the solr-start goal (as shown in the pom), to ensure that any instances of solr that may already be running on that port are ended.

### Benefits

* Checks status codes of commands it runs.
* Actively provides feedback in the maven logs about its status.
* Will halt the build if Solr does not start correctly, rather than letting integration tests begin.
* Customizable via configuration options.
* Includes its own integration tests to verify functionality.

### Possible Drawbacks

* Built-in Integration tests require changing the system's .m2/settings.xml.
* Built-in Integration tests add several minutes to a build.

### Verification Process

* Functionality tested on both Windows and Linux environments.
* Created tests asserting that the plugin acts as expected.

### Review
@oconnormi

### CODE REVIEW

**Comment by paouelle** [comment]
❓ Why are we not changing the way we start itests instead to be more like the component tests where it starts the distro using the ddf start script instead of karaf? 

This would have the advantage of:
- starting Solr the right way based on how the distro is configured.
- allowing us to dynamically choose a Solr port as opposed to hardcoding one.
- also make sure that anything else which needs to be started for DDF is properly handled using the startup script we already maintained

File: solr-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>solr-plugin</artifactId>
+    <packaging>maven-plugin</packaging>
+    <name>solr-plugin Maven Mojo</name>
+    <url>http://maven.apache.org</url>
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.maven</groupId>
+            <artifactId>maven-plugin-api</artifactId>
+            <version>3.6.0</version>
```
Comment (paouelle) [inline solr-plugin/pom.xml:17]: ❗️ Versions should be defined in the top pom using properties. Also add these dependencies in its dependency management section along with the versions referencing those properties. Here, only reference the artifact without the version.

File: solr-plugin/pom.xml [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.8.1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.maven.shared</groupId>
+            <artifactId>maven-invoker</artifactId>
+            <version>3.0.1</version>
+        </dependency>
+        <dependency>
```
Comment (paouelle) [inline solr-plugin/pom.xml:44]: ❗️ codice-maven should not have any dependencies to DDF as this would create a circular dependency.

**Review by paouelle — CHANGES_REQUESTED** [review]
We cannot introduce dependencies to DDF in codice-maven as it would introduce circular dependencies. This is the wrong place for such plugin.

I am stopping my review for now as we need to relocate this plugin elsewhere if we are to keep it. Since this plugins has a dependency to DDF and DDF will have a dependency to that plugin, it should be part of that repo.

**Comment by aperullo** [comment]
No longer needed.

**Pull request closed without merging**